File: Depth-First Search/Graph/LET_no1463_Cherry Pickup II.py
class Solution:
    def cherryPickup(self, grid: List[List[int]]) -> int:
        return self.dfs(grid, dict(), 0, 0, 0, len(grid[0]) - 1)
    
    def dfs(self, grid, memo, i1, j1, i2, j2):
        if (i1, j1, i2, j2) in memo:
            return memo[(i1, j1, i2, j2)]
        
        res = 0
        for d1 in (-1, 0, 1):
            for d2 in (-1, 0, 1):
                x1, x2, y1, y2 = i1 + 1, i2 + 1, j1 + d1, j2 + d2
                if x1 < len(grid) and 0 <= y1 < len(grid[0]) and 0 <= y2 < len(grid[0]):
                    res = max(res, self.dfs(grid, memo, x1, y1, x2, y2))
        
        if j1 == j2:
            res += grid[i1][j1]
        else:
            res += (grid[i1][j1] + grid[i2][j2])
        memo[(i1, j1, i2, j2)] = res
        return res

File: Depth-First Search/Graph/LET_no1463_Cherry Pickup II.cpp
class Solution {
public:
    int dp[70][70][70];
    int cherryPickup(vector<vector<int>>& grid) {
        memset(dp, -1, sizeof(dp));
        int m = grid.size(), n = grid[0].size();
        return dfs(grid, m, n, 0, 0, n - 1);
    }
    
    int dfs(vector<vector<int>>& grid, int m, int n, int r, int c1, int c2) {
        if (dp[r][c1][c2] != -1)
            return dp[r][c1][c2];
        
        int res = 0;
        for (int d1 = -1; d1 <= 1; d1++) {
            for (int d2 = -1; d2 <= 1; d2++) {
                int c3 = c1 + d1, c4 = c2 + d2, r2 = r + 1;
                if (r2 < m && 0 <= c3 && c3 < n && 0 <= c4 && c4 < n)
                    res = max(res, dfs(grid, m, n, r2, c3, c4));
            }
        }
        
        res += c1 == c2 ? grid[r][c1] : grid[r][c1] + grid[r][c2];
        dp[r][c1][c2] = res;
        return res;
    }
};